### PR TITLE
Normalization changes

### PR DESCRIPTION
**Pull request by roystgnr**
Autoscale by simulation data only, to better match the MATLAB code.

Gracefully handle the "we have nothing to scale by if we leave out the experimental data" cases; even with a dummy variable where we don't care about the scaling, NaNs spread like weeds.

### CODE REVIEW

**Comment by roystgnr** [comment]
Pretty sure this PR doesn't *actually* break test_uqGaussianVectorRVClass on clang.

Stupid stochastic tests.  Are we not using a fixed PRNG seed in the test suite?  That wouldn't fix "new algorithm calls for random numbers in a different order" false positives, but it would at least fix "old algorithm gets different random numbers this time" false positives.

File: src/gp/src/GPMSAOptions.C [hunk 1]
```diff
@@ -554,9 +554,12 @@ GPMSAOptions::set_final_scaling
         min_max_update(minScenario, maxScenario,
                        (*m_simulationScenarios[i]));
 
+      // To match Matlab code, only use simulation data for scaling.
```
Comment (dmcdougall) [inline src/gp/src/GPMSAOptions.C]: Just delete this code instead of commenting it out.

File: src/gp/src/GPMSAOptions.C [hunk 2]
```diff
@@ -634,9 +637,12 @@ GPMSAOptions::set_final_scaling
         mean_var_update(n, meanScenario, varScenario,
                         *m_simulationScenarios[i]);
 
+      // To match Matlab code, only use simulation data for scaling.
```
Comment (dmcdougall) [inline src/gp/src/GPMSAOptions.C]: Ditto.

**Review by dmcdougall — CHANGES_REQUESTED** [review]

File: src/gp/src/GPMSAOptions.C (hunk 1 above)
Comment (roystgnr) [reply src/gp/src/GPMSAOptions.C]: I failed to come up with a decent middle ground there, huh?  I was trying to avoid both the confusing Scylla of "just delete the code, and leave future generations with no indication of why we're ignoring half of our input data" and the passive-aggressive Charybdis of "// Only use simulation data for scaling.  This is necessary to match Matlab code output, not for any sane reason."  ;-)

Would it be reasonable for me to leave the code in, but wrapped in an "autoscale_including_experiment_data" option (which defaults to false, if you like)?  Or should I move the documentation to the autoscale method declarations?  Or both?  (I'm now leaning toward "both".)

I'm seriously insisting on *some* sort of documentation to exist regarding why we're making this choice.  "Why are we normalizing numbers to a (min,max) range which will leave numbers outside that range?" is a question with a defensible answer, but "Why are we doing so without telling users?" is not.

**Comment by brianw525** [comment]
The Gaussian process surrogate used for calibration in GPMSA is built using the simulation data only. Therefore, it is really only valid on the convex hull of the simulation inputs (scenario and calibration). Anything outside of that is an extrapolation, and we should never extrapolate with surrogates as a general rule. The choice of simulation inputs only for normalization is therefore motivated by the fact that they constitute the training data for the surrogate and the interpretation of correlation length parameters is based on the design region chosen for generating the simulation data.

No user should be calibrating to experimental scenarios that fall outside of the range of scenarios determined by the simulation data. Since we are normalizing with respect to the simulation inputs, I would suggest that QUESO (or Dakota) throws an error (or warning) if any of the experimental scenario variables take a normalized value greater than 1 or less than 0. The practitioner should ensure that the range of scenarios passed to the simulator covers the range of experimental scenarios used to calibrate the model parameters.

**Comment by roystgnr** [comment]
We have more than one gaussian process in play here, though, right?  The GP modeling the simulator is calibrated using only simulator data, but the GP modeling the discrepancy must be calibrated using experiment data points.  Is it really obvious that normalizing the former to a too-wide data range would be worse than normalizing the latter to a too-narrow range?

I'll put this in docs, regardless.  I like your idea about a warning, too, and I can add a bool+getter in case Dakota wants to turn that warning into an error.

**Comment by brianw525** [comment]
Yes we do, however both the GP modeling the simulator AND the GP modeling the discrepancy are involved in the calibration. Therefore if we have experimental scenarios outside the range of the simulator scenarios, we are extrapolating regardless of what normalization we use and this should not be allowed as a matter of principle.

Bottom line is we need to be normalizing scenarios (and calibration parameters of course) with respect to the simulator design only. This is the convention utilized in the Matlab code and there is no justified reason for changing it. It needs to be done the same way in the Matlab code and in QUESO to make verification feasible.

File: src/gp/src/GPMSAOptions.C [hunk 3]
```diff
@@ -550,16 +564,18 @@ GPMSAOptions::set_final_scaling
...
-/*
+      // But check the experimental data, and yell at the user if it
+      // falls outside the simulation data range, because they
+      // probably don't want to be extrapolating with their GP.
       for (unsigned int i=0; i < m_experimentScenarios.size(); ++i)
         min_max_update(minScenario, maxScenario,
-                       (*m_experimentScenarios[i]));
-*/
+                       (*m_experimentScenarios[i]),
+                       /*warn_on_update =*/ "scenario parameter");
```
Comment (dmcdougall) [inline src/gp/src/GPMSAOptions.C]: If `min_max_update` is still being called on the experiment scenarios, then isn't the min and max including the experiment scenarios in the normalization?

Comment (dmcdougall) [inline src/gp/src/GPMSAOptions.C]: If min_max_update is still being called on the experiment scenarios, then isn't the min and max including the experiment scenarios in the normalization?

**Review by dmcdougall — CHANGES_REQUESTED** [review]

File: src/gp/src/GPMSAOptions.C (hunk 3 above)
Comment (dmcdougall) [reply src/gp/src/GPMSAOptions.C]: Sorry, that comment is horribly worded.  Let me try again.

If `min_max_update` is still being called on the experiment scenarios, then we're including the experiment scenarios in the normalization.

Comment (roystgnr) [reply src/gp/src/GPMSAOptions.C]: Yeah, sorry; that last commit should fix it.

**Review by dmcdougall — APPROVED** [review]
